### PR TITLE
Improve encapsulation of components used in FetchBatchAccumulator

### PR DESCRIPTION
**Pull request by mfussenegger**
### Improve encapsulation of components used in FetchBatchAccumulator

- Renames `FetchProjectorContext` → `ReaderBuckets`. Is now used only
for partitioning of the `_fetchId`s by `_readerId` into a `_docId` +
  later carry the result for the fetched values for the doc-ids.

- `FetchRowInputSymbolVisitor` leaked a lot of the workings into the
`FetchBatchAccumulator`. This is now encapsulated via `FetchRows`.

### Replace AsyncOperationBatchIterator with AsyncFlatMapBatchIterator 

The AsyncOperationBatchIterator was a combination of two things:

- Building chunks (batches)
- Async invocation of an operation, and generating rows based on the
result of this operation.

This introduces a new `AsyncFlatMapBatchIterator` that is similar to
`FlatMapBatchIterator`, but takes a mapper that returns a future.

The "building chunks" part is separated by using
`BatchIterators.partition`.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: dex/src/main/java/io/crate/data/AsyncFlatMapper.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package io.crate.data;
+
+import java.util.Iterator;
+import java.util.concurrent.CompletableFuture;
+
+public interface AsyncFlatMapper<I, O> extends AutoCloseable {
+
+    CompletableFuture<? extends Iterator<O>> apply(I item, boolean isLastCall);
```
Comment (mkleen) [inline dex/src/main/java/io/crate/data/AsyncFlatMapper.java]: It is not so easy to see the `flatmap` here, seems like this only provides the map operation and the rest happens in `AsyncFlatMapBatchIterator`. However this makes all sense. Maybe a few lines of documentation/comments would be good.

File: sql/src/main/java/io/crate/execution/engine/fetch/FetchMapper.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import io.crate.concurrent.CompletableFutures;
+import io.crate.data.AsyncFlatMapper;
+import io.crate.data.Bucket;
+import io.crate.data.Row;
+
+public class FetchMapper implements AsyncFlatMapper<ReaderBuckets, Row> {
```
Comment (mkleen) [inline sql/src/main/java/io/crate/execution/engine/fetch/FetchMapper.java]: Very nice!

File: dex/src/main/java/io/crate/data/AsyncFlatMapper.java (hunk 1 above)
Comment (mfussenegger) [reply dex/src/main/java/io/crate/data/AsyncFlatMapper.java]: Yep, this is modelled after the stream API:

```
flatMap(Function<? super RewriteToQueryThenFetch,? extends Stream<? extends R>> mapper) : Stream<R>
```

The function that does the mapping is called `mapper`.

I will add some more docs to clarify that.

Comment (mfussenegger) [reply dex/src/main/java/io/crate/data/AsyncFlatMapper.java]: I tried to add something, but it read as a repitition of the signature, so I'll keep it as it is for now.

**Pull request merged**